tests/autorestart_test/test_container_autorestart.py

```diff
@@ -0,0 +1,295 @@
+    file_content = duthost.shell("docker exec {} bash -c '[ -f /etc/supervisor/critical_processes ] \
+        && cat /etc/supervisor/critical_processes'".format(container_name), module_ignore_errors=True)
+    for line in file_content["stdout_lines"]:
+        line_info = line.strip(' \n').split(':')
+        if len(line_info) != 2:
+            pytest_assert(False, "Syntax of the line {} in critical_processes \
+                file is incorrect.".format(line))
+        
+        identifier_key = line_info[0].strip()
+        identifier_value = line_info[1].strip()
+        if identifier_key == "group" and identifier_value:
+            critical_group_list.append(identifier_value)
+        elif identifier_key == "program" and identifier_value:
+            critical_process_list.append(identifier_value)
+        else:
+            pytest_assert(False, "Syntax of the line {} in critical_processes \
+                file is incorrect.".format(line))
+ 
+    return critical_group_list, critical_process_list
+
+def get_group_program_info(duthost, container_name, critical_group):
+    """
+    @summary: Get critical program names, running status and their pids by analyzing the command
+              output of "docker exec <container_name> supervisorctl"
+    @return: Critical program names and their pids
+    """
+    group_program_info = defaultdict(list)
+    program_name = None
+    program_status = None
+    program_pid = -1
+
+    program_list = duthost.shell("docker exec {} supervisorctl".format(container_name))
+    for program_info in program_list["stdout_lines"]:
+        if program_info.find(critical_group) != -1:
+            program_name = program_info.split()[0].split(':')[1].strip(' \n')
+            program_status = program_info.split()[1].strip()
+            if program_status in ["EXITED", "STOPPED", "STARTING"]:
+                program_pid = -1
+            else:
+                program_pid = program_info.split()[3].strip(' ,')
+
+            group_program_info[program_name].append(program_status)
+            group_program_info[program_name].append(program_pid)
+            
+    return group_program_info
+
+def get_program_info(duthost, container_name, program_name):
+    """
+    @summary: Get program running status and pid by analyzing the command
+              output of "docker exec <container_name> supervisorctl"
+    @return:  Program running status and its pid
+    """
+    program_status = None
+    program_pid = -1
+
+    program_list = duthost.shell("docker exec {} supervisorctl".format(container_name))
+    for program_info in program_list["stdout_lines"]:
+        if program_info.find(program_name) != -1:
+            program_status = program_info.split()[1].strip()
+            if program_status == "RUNNING":
+                program_pid = int(program_info.split()[3].strip(' ,'))
+            break
+
+    logging.info("Found program {} in the {} state with pid {}".format(program_name, program_status, program_pid))
```

Comment:
What happens if you don't find `program_name` in the supervisorctl output?

Reply:
If `program_name` is not found in the supervisorctl output, then the function `get_program_info(...)` will return program_status which is `None`. At the line 147, the function `kill_process_by_pid(...)` will throw an error message to indicate the process is not found in that container.

Reply:
Resolved via refactor of `get_program_info`